compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -944,8 +944,31 @@ public void visit(BLangRecordLiteral recordLiteral) {
         }
     }
 
-    private List<BType> getRecordCompatibleType(BType bType, BLangRecordLiteral recordLiteral) {
+    private boolean isMappingConstructorCompatibleType(BType type) {
+        return type.tag == TypeTags.RECORD || type.tag == TypeTags.MAP || type.tag == TypeTags.JSON;
+    }
+
+    private void reportIncompatibleMappingConstructorError(BLangRecordLiteral mappingConstructorExpr, BType expType) {
+        if (expType.tag != TypeTags.UNION) {
+            dlog.error(mappingConstructorExpr.pos, DiagnosticCode.MAPPING_CONSTRUCTOR_COMPATIBLE_TYPE_NOT_FOUND,
+                       expType);
+            return;
+        }
+
+        BUnionType unionType = (BUnionType) expType;
+        Set<BType> memberTypes = unionType.getMemberTypes();
+
+        for (BType bType : memberTypes) {
+            if (isMappingConstructorCompatibleType(bType)) {
+                dlog.error(mappingConstructorExpr.pos, DiagnosticCode.INCOMPATIBLE_MAPPING_CONSTRUCTOR, unionType);
+                return;
```

Comment:
should we not print all the incompatible member types at once?

Reply:
This is just to isolate instances where the union has a type with which we can use record literals (i.e., records, maps an JSON), but for this particular case, the given record literal isn't compatible with that type.

Reply:
How about cases like 

`Foo|Bar x = {};`

where both `Foo` and `Bar` are records but `{}` is not valid for either of them?

That still reaches here right? If it does IMO, we should mention both of them.

Reply:
The error message says the mapping constructor is not valid for the union. It doesn't mention individual types. For example, for the above case you mentioned:
```
incompatible mapping constructor expression for type '(Foo|Bar)'
```

Reply:
ah yeah, missed that we were using `unionType`, sorry!